### PR TITLE
Tau id common function

### PR DESCRIPTION
**Pull request by GaelTouquet**

### CODE REVIEW

**Comment by GaelTouquet** [comment]
Goes with https://github.com/cbernet/cmgtools-lite/pull/31

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 1]
```diff
@@ -35,7 +37,7 @@ def mva_score(self, norm=False):
 
     def mva_passes(self, working_point):
         '''returns True if the tau passes the given working point of the isolation mva'''
-        return tau_mvaid.passes(self, working_point)
+        return self._mvaid_normscore > tau_mvaid.threshold(self.pt(), working_point)
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: need to test the existence of the cache here and not outside, as this function can be called directly by the user.
in the same way, mva_score must test the cache and set it if necessary. 
there should be no way for the user to hit a None cache

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 2]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
         return ( - (vtx.x()-vertex.position().x()) *  p4.y()
                  + (vtx.y()-vertex.position().y()) *  p4.x() ) /  p4.pt()
 
+    def countWP(self,name, allWP=False):
+        '''Returns the number of Working Points 
+        that are passed for given isolation name.
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: too vague, the user cannot understand what is the returned value from this docstring.
should give the list of working points and the returned value (maybe with an example)
should explain what happens when the wp state is read from mini aod
should explain what name and allWP should be and what they do.
I don't understand why allWP is necessary

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 3]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+    def countWP(self,name, allWP=False):
+        '''Returns the number of Working Points 
+        that are passed for given isolation name. 
+        '''
+        WPs = ['VVTight', 'VTight', 'Tight', 'Medium',
+               'Loose', 'VLoose', 'VVLoose', ]
+        testname = 'by{}'+name[2:]
+        n_WP = 0
+        for WP in WPs:
+
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: rm empty line. 
in general empty lines are not needed in python, keep them at the minimum.

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 4]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+        n_WP = 0
+        for WP in WPs:
+            
+            # if from heppy-implemented isolation (known list of WPs)
+            if allWP:
+                if not self.tauID(testname.format(WP)):
+                    n_WP += 1 # in this case n_WP is the number of not passed WP
+                else:
+                    return len(WPs) - n_WP #stops counting when reached a passed WP
+
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: rm empty line

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 5]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+            # else all WPs will be tested from MINIAOD
+            elif self.physObj.isTauIDAvailable(testname.format(WP)):
+                if self.tauID(testname.format(WP)):
+                    n_WP += 1 # in this case n_WP is the number of passed WP
+
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: rm empty line

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 6]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
         return ( - (vtx.x()-vertex.position().x()) *  p4.y()
                  + (vtx.y()-vertex.position().y()) *  p4.x() ) /  p4.pt()
 
+    def countWP(self,name, allWP=False):
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: space missing before name

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 7]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
-            return self.mva_passes("Eff70")
-        elif name == 'byTightIsolationMVArun2017v2DBoldDMwLT2017':
-            return self.mva_passes("Eff60")
-        elif name == 'byVTightIsolationMVArun2017v2DBoldDMwLT2017':
-            return self.mva_passes("Eff50")
-        elif name == 'byVVTightIsolationMVArun2017v2DBoldDMwLT2017':
-            return self.mva_passes("Eff40")
+        if 'IsolationMVArun2017v2DBoldDMwLT' in name:
+            #if raw score
+            if name == 'byIsolationMVArun2017v2DBoldDMwLTraw2017':
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: remove, will test cache in mva_score and evaluate there if necessary

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 8]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+                if self._mvaid_score is None:
+                    self._mvaid_score = self.mva_score()
+                return self._mvaid_score
+
+            #if WPs
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: same, will test cache in mva_passes

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 9]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+            if self._mvaid_normscore is None:
+                self._mvaid_normscore = self.mva_score(norm=True)
+
+            if name == 'byVVLooseIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VVLoose") else 0.
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: just call mva_passes directly (everywhere)

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 10]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+            elif name == 'byVLooseIsolationMVArun2017v2DBoldDMwLT2017':
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: actually,  should add a dictionary mapping the working point name and our internal name, 
and have the call at a single place: 

```python
wp_map = dict(
...
 byLooseIsolationMVArun2017v2DBoldDMwLT2017 = 'Loose'
...
)

if name in wp_map:
    return self.mva_passes( wp_map[name])
else:
    return self.physObj.tauID(name)
```

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 11]
```diff
@@ -46,27 +48,68 @@ def dxy_approx(self, vertex=None):
...
+                return 1. if self.mva_passes("Medium") else 0.
+            elif name == 'byTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("Tight") else 0.
+            elif name == 'byVTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VTight") else 0.
+            elif name == 'byVVTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VVTight") else 0.
+
+            #if numbering scheme
+            if ('byIsolation' in name) and ('raw' not in name):
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: this is the case of byLooseIsolationMVArun2017v2DBoldDMwLT2017.... one can easily be mislead by the fact that many elses are missing. 
do not rely on the return to make sure that we don't hit a given piece of code at the end. instead, add explicit elses, see above.

**Review by cbernet — CHANGES_REQUESTED** [review]

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 12]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
         return ( - (vtx.x()-vertex.position().x()) *  p4.y()
                  + (vtx.y()-vertex.position().y()) *  p4.x() ) /  p4.pt()
 
+    def countWP(self,name, allWP=False):
+        '''Returns the number of Working Points 
+        that are passed for given isolation name. 
+        '''
+        WPs = ['VVTight', 'VTight', 'Tight', 'Medium',
+               'Loose', 'VLoose', 'VVLoose', ]
+        testname = 'by{}'+name[2:]
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: adding string is very inefficient. join them with join

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 13]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+            if allWP:
+                if not self.tauID(testname.format(WP)):
+                    n_WP += 1 # in this case n_WP is the number of not passed WP
+                else:
+                    return len(WPs) - n_WP #stops counting when reached a passed WP
+
+            # else all WPs will be tested from MINIAOD
+            elif self.physObj.isTauIDAvailable(testname.format(WP)):
+                if self.tauID(testname.format(WP)):
+                    n_WP += 1 # in this case n_WP is the number of passed WP
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: do you mean that the returned value is different for the two modes? this cannot be.

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 14]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+        '''
+        WPs = ['VVTight', 'VTight', 'Tight', 'Medium',
+               'Loose', 'VLoose', 'VVLoose', ]
+        testname = 'by{}'+name[2:]
+        n_WP = 0
+        for WP in WPs:
+            
+            # if from heppy-implemented isolation (known list of WPs)
+            if allWP:
+                if not self.tauID(testname.format(WP)):
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: the fact that tauID calls countWP and countWP calls tau ID is very confusing, and should be explained.

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 15]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+        '''Returns heppy redefined score if fitting one 
+        of implemented names, else calls the wrapped tau 
+        tauID(name) function.
+        '''
+        if 'IsolationMVArun2017v2DBoldDMwLT' in name:
+            #if raw score
+            if name == 'byIsolationMVArun2017v2DBoldDMwLTraw2017':
+                if self._mvaid_score is None:
+                    self._mvaid_score = self.mva_score()
+                return self._mvaid_score
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: should be under the if

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 16]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+
+            #if WPs
+            if self._mvaid_normscore is None:
+                self._mvaid_normscore = self.mva_score(norm=True)
+
+            if name == 'byVVLooseIsolationMVArun2017v2DBoldDMwLT2017':
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: should be elif

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 17]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+            #if raw score
+            if name == 'byIsolationMVArun2017v2DBoldDMwLTraw2017':
+                if self._mvaid_score is None:
+                    self._mvaid_score = self.mva_score()
+                return self._mvaid_score
+
+            #if WPs
+            if self._mvaid_normscore is None:
+                self._mvaid_normscore = self.mva_score(norm=True)
+
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: an else should be there. this happens only when not in raw mode and not in auto WP mode

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 18]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+            elif name == 'byMediumIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("Medium") else 0.
+            elif name == 'byTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("Tight") else 0.
+            elif name == 'byVTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VTight") else 0.
+            elif name == 'byVVTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VVTight") else 0.
+
+            #if numbering scheme
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: I don't understand this part at all.

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 19]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+        '''
+        if 'IsolationMVArun2017v2DBoldDMwLT' in name:
+            #if raw score
+            if name == 'byIsolationMVArun2017v2DBoldDMwLTraw2017':
+                if self._mvaid_score is None:
+                    self._mvaid_score = self.mva_score()
+                return self._mvaid_score
+
+            #if WPs
+            if self._mvaid_normscore is None:
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: I don't understand what this is doing here, not protected by any if. 
I would expect the call to countWP to be here.

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 20]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+            elif name == 'byVTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VTight") else 0.
+            elif name == 'byVVTightIsolationMVArun2017v2DBoldDMwLT2017':
+                return 1. if self.mva_passes("VVTight") else 0.
+
+            #if numbering scheme
+            if ('byIsolation' in name) and ('raw' not in name):
+                return self.countWP(name, allWP=True)
+
+        if ('byIsolation' in name) and ('raw' not in name):
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: but that at the beginning, as previously stated

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 21]
```diff
@@ -44,6 +48,70 @@ def dxy_approx(self, vertex=None):
...
+            return self.countWP(name)
+
+        return self.physObj.tauID(name)
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: there should be another else protecting this return

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 22]
```diff
@@ -32,12 +32,19 @@ def mvaId(self):
...
+            self._mvaid_score = tau_mvaid.score(self)
+        return self._mvaid_score
 
     def mva_passes(self, working_point):
         '''returns True if the tau passes the given working point of the isolation mva'''
-        return self._mvaid_normscore > tau_mvaid.threshold(self.pt(), working_point)
+        if self._mvaid_normscore is None:
+            self.mva_score(norm=True)
+        passes = self._mvaid_normscore > tau_mvaid.threshold(self.pt(), working_point)
+        return 1. if passes else 0.
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: add comment

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 23]
```diff
@@ -48,69 +55,62 @@ def dxy_approx(self, vertex=None):
...
+        For example if a tau only passes VLoose, Loose,
+        and Medium Working Points this will return 3.
+
+        Uses self.tauID() to evaluate the WPs.
         '''
         WPs = ['VVTight', 'VTight', 'Tight', 'Medium',
-               'Loose', 'VLoose', 'VVLoose', ]
-        testname = 'by{}'+name[2:]
+               'Loose', 'VLoose', 'VVLoose']
+        testname = ''.join(['by{}',name[2:]])
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: do a replace

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 24]
```diff
@@ -48,69 +55,62 @@ def dxy_approx(self, vertex=None):
...
+        #using numbering scheme
         if ('byIsolation' in name) and ('raw' not in name):
             return self.countWP(name)
-
-        return self.physObj.tauID(name)
+        # search in MINIAOD
+        if self.physObj.isTauIDAvailable(name):
+            return self.physObj.tauID(name)
+        # name not available
+        if verbose:
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: else

Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: raise ValueError with this message if not found

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 25]
```diff
@@ -48,69 +55,62 @@ def dxy_approx(self, vertex=None):
...
-            elif self.physObj.isTauIDAvailable(testname.format(WP)):
-                if self.tauID(testname.format(WP)):
-                    n_WP += 1 # in this case n_WP is the number of passed WP
-
+            if self.tauID(testname.format(WP), verbose=False):
+                n_WP += 1
         return n_WP
 
-    def tauID(self, name):
+    def tauID(self, name, verbose=True):
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: remove verbose

File: PhysicsTools/Heppy/python/physicsobjects/Tau.py [hunk 26]
```diff
@@ -48,69 +55,62 @@ def dxy_approx(self, vertex=None):
...
-                    n_WP += 1 # in this case n_WP is the number of not passed WP
-                else:
-                    return len(WPs) - n_WP #stops counting when reached a passed WP
-
-            # else all WPs will be tested from MINIAOD
-            elif self.physObj.isTauIDAvailable(testname.format(WP)):
-                if self.tauID(testname.format(WP)):
-                    n_WP += 1 # in this case n_WP is the number of passed WP
-
+            if self.tauID(testname.format(WP), verbose=False):
```
Comment (cbernet) [inline PhysicsTools/Heppy/python/physicsobjects/Tau.py]: catch valueerror to deal with missing wps in miniaod

**Review by cbernet — CHANGES_REQUESTED** [review]

**Review by cbernet — APPROVED** [review]

**Pull request merged**